app/src/main/java/com/example/practice/MainActivityViewModel.kt

```diff
@@ -35,7 +35,7 @@ class MainActivityViewModel : ViewModel() {
     fun fetchUserList(excludingUserId: String = ""){
 
         // Call Network Api to get user data
-        RetrofitAPI().fetchUsersList(excludingUserId, {
+        API.create().fetchUsersList(excludingUserId, {
```

Comment:
The initialisation `API.create()` is executed each time this function is invoked. Meaning each network fetch is handled by a different instance of `API` implementation. 

1. What is the advantage and disadvantage of this approach?
2. What happens in case of multiple simultaneous calls (e.g. if we call this function from a *Refresh* button action)?
3. What makes each instance stay in memory after the function scope has exited? 

- [ ] Are all these questions clear? If not, comment your question below

Reply:
@nets-dev

Q1] What is the advantage and disadvantage of this approach?
**Ans** :- Advantage - 1. It hides the business logic of the component.
        			      2. It is loosely coupled with components. So that we can easily replace our business logic with new approach.
			              3. Better to maintain code structure.
          Disadvantage -   1. We have to take care that API.create() should return the same instance for each network call otherwise garbage memory increases.

Q2] What happens in case of multiple simultaneous calls (e.g. if we call this function from a Refresh button action)?
**Ans** :- RetrofitAPI instance will be get created multiple times and garbage collection will increase.

Q3] What makes each instance stay in memory after the function scope has exited?
**Ans** : - Yes. Instances stay in memory. It is because, in API.create() function we are returning a new instance of RetrofitAPI class every time.

Q4] Is it possible the application can struggle to garbage collect these instances? If yes, in what scenario? If no, how can we be sure that will not happen? (Give straightforward and concrete scenarios/reasons)
**Ans** :- Yes, it is possible applications can struggle to garbage collect these instances. Whenever we make a new api request, a new RetrofitAPI instance is created which leads to an increase in garbage memory. To avoid it we can return the single instance of RetrofitAPI from API.create() function.

Reply:
@lnt-nishantdande Thank you for the answers! I think there is a misunderstanding on question 3 so I'll explain further. Local variables are freed when a function scope exits, e.g. 
`fun printHelloWorld() {
    val helloWorld: String = "Hello World!"
    print(helloWorld)
}`
The local property `helloWorld` is freed at the end of the function scope, in other words, after the closing bracket - **}**. That means, the local property obtained from `API.create()` should normally be freed as soon as `fetchUserList` function reaches end of function scope at line 62. 
So the question is, what makes this instance stay in memory way beyond the function scope and execute a callback upon networking response. 
Is the question now clear?

Reply:
@nets-dev Thank you for explaining me the Q3 in detail.
According to me, as we are calling create function statically means create function is placed in companion object { } and hence instance stay in memory till application lifecycle.